HCP Terraform Docs (#1664)
hashicorp/web-unified-docs: content/terraform-docs-common/docs/cloud-docs/registry/test/dynamic-credentials/vault.mdx

---
page_title: Use dynamic credentials with Vault in module testing
description: >-
  Configure HashiCorp Vault JWT auth to trust HCP Terraform's OIDC tokens for module test runs,
  enabling dynamic secret generation without static credentials.
---

# Configure dynamic credentials for Vault in module testing

@include 'tfc-package-callouts/registry-test-dynamic-credentials.mdx'

You can use HCP Terraform's native OpenID Connect integration with HashiCorp Vault to get dynamic secrets for module test runs. This approach uses Vault's JWT/OIDC auth method to authenticate and generate short-lived credentials or secrets without storing static Vault tokens.

## Requirements

You need the following to configure dynamic credentials for Vault in module testing:

> - A HashiCorp Vault instance (Vault 1.8.0 or later) with permissions to configure auth methods and policies

Reviewer: ```suggestion
- A HashiCorp Vault instance version 1.8.0 or later with permissions to configure auth methods and policies
```

- An HCP Terraform organization with module testing enabled
- A private registry module with [testing configured](/terraform/cloud-docs/registry/test)
- Permissions in HCP Terraform to modify registry module test settings
- The `vault` CLI (optional, for command-line configuration)

## Configure Vault

### Enable the JWT auth method

Enable the JWT auth method in Vault if it's not already enabled:

```bash
vault auth enable jwt
```

If you want to use a different path (e.g., for organization isolation):

```bash
vault auth enable -path=terraform-module-tests jwt
```

### Configure the JWT auth method

Configure the auth method to trust HCP Terraform's OIDC issuer:

```bash
vault write auth/jwt/config \
  oidc_discovery_url="https://app.terraform.io" \
  bound_issuer="https://app.terraform.io"
```

> For Terraform Enterprise, replace `app.terraform.io` with your TFE hostname:

Reviewer: ```suggestion
For Terraform Enterprise, replace `app.terraform.io` with your Terraform Enterprise hostname:
```

}
EOF
```

Adjust the policy based on your module's testing requirements. Follow the principle of least privilege.

### Create a JWT auth role

Create a role that binds the Vault policy to HCP Terraform module test runs:

```bash
vault write auth/jwt/role/module-tests \
  role_type="jwt" \
  bound_audiences="vault.workload.identity" \
  bound_claims_type="glob" \
  bound_claims="/terraform_run_phase=plan,terraform_organization_name=my-org" \
  bound_subject="organization:my-org:module:*:operation:test_run" \
  claim_mappings="/terraform_run_id=run_id,terraform_organization_name=organization_name,terraform_run_phase=run_phase" \
  user_claim="terraform_run_id" \
  token_policies="module-test-policy" \
  token_ttl="1h"
```

This configuration:
- Requires the audience to be `vault.workload.identity`
- Only accepts tokens where `terraform_run_phase` is `plan` (test runs always use the plan phase)
- Uses the `bound_subject` to match the module test subject claim format (`organization:{ORG_NAME}:module:{MODULE_NAME}:operation:test_run`)
- Maps token claims to Vault identity metadata
- Grants the `module-test-policy` to authenticated requests
- Issues Vault tokens valid for 1 hour

> ### More specific bound claims

Reviewer: ```suggestion
### Refine bound claims
```
I had a question about the same heading elsewhere.

> You can create more granular access controls using bound claims:

Reviewer: ```suggestion
You can create more granular access controls using bound claims. 
```

**Allow specific module using subject claim:**
```bash
vault write auth/jwt/role/vpc-module-tests \
  role_type="jwt" \
  bound_audiences="vault.workload.identity" \
  bound_subject="organization:my-org:module:terraform-aws-vpc:operation:test_run" \
  bound_claims="/terraform_run_phase=plan" \
  token_policies="vpc-test-policy" \
  token_ttl="30m"
```

**Allow specific organization:**
```bash
vault write auth/jwt/role/org-module-tests \
  role_type="jwt" \
  bound_audiences="vault.workload.identity" \
  bound_claims="/terraform_run_phase=plan,terraform_organization_name=my-org" \
  bound_subject="organization:my-org:module:*:operation:test_run" \
  token_policies="module-test-policy" \
  token_ttl="1h"
```

> **Use subject claim pattern for all modules:**

Reviewer: ```suggestion
The following command uses the subject claim pattern to create a role that binds the claim to all modules  for all modules:
```

```hcl
data "vault_generic_secret" "test_config" {
  path = "secret/data/module-tests/config"
}

# Use the secret in your test
resource "aws_instance" "test" {
  ami           = data.vault_generic_secret.test_config.data["ami_id"]
  instance_type = "t2.micro"
}
```

### Generating dynamic database credentials

```hcl
data "vault_database_secret_backend_connection" "db" {
  backend = "database"
  name    = "readonly"
}

# Use the generated credentials
provider "postgresql" {
  host     = data.vault_database_secret_backend_connection.db.data["host"]
  username = data.vault_database_secret_backend_connection.db.data["username"]
  password = data.vault_database_secret_backend_connection.db.data["password"]
}
```

### Generating PKI certificates

```hcl
resource "vault_pki_secret_backend_cert" "test_cert" {
  backend     = "pki"
  role        = "module-tests"
  common_name = "test.example.com"
  ttl         = "24h"
}

# Use the certificate in your test
resource "aws_acm_certificate" "test" {
  private_key      = vault_pki_secret_backend_cert.test_cert.private_key
  certificate_body = vault_pki_secret_backend_cert.test_cert.certificate
}
```

## Policy best practices

> When creating Vault policies for module tests:

Reviewer: ```suggestion
We recommend adhering to the following best practices when creating Vault policies for module tests:
```

   ```bash
   vault write auth/jwt/role/module-tests \
     token_ttl="30m" \
     token_max_ttl="1h"
   ```

5. **Audit access**: Enable Vault audit logging to track secret access
   ```bash
   vault audit enable file file_path=/var/log/vault/audit.log
   ```

## Verify the configuration

After configuring both Vault and HCP Terraform, verify the setup by running a test:

1. Navigate to your module in HCP Terraform
2. Go to the **Tests** tab
3. Click **Start Test Run** or trigger a test through your VCS
4. Monitor the test run logs for successful Vault authentication

If authentication succeeds, you'll see log output indicating the Vault provider authenticated using the OIDC token. If it fails, check the [troubleshooting](#troubleshooting) section.

## Troubleshooting

### Error: "Permission denied"

The JWT role denied the token. Check:

1. The JWT auth method is enabled at the correct path
2. The role exists and is configured correctly
3. The `bound_audiences` includes `vault.workload.identity`
4. The `bound_claims` match your test run's token claims
5. The subject claim matches the `bound_subject` (if configured)

### Error: "Invalid JWT"

> The token validation failed. Verify:

Reviewer: ```suggestion
The token validation failed. Verify the following conditions:
```

1. The `oidc_discovery_url` in Vault points to HCP Terraform's OIDC endpoint
2. Your Vault instance can reach the OIDC discovery URL
3. The `bound_issuer` matches the token's issuer exactly
4. Vault is using version 1.8.0 or later

### Error: "Vault address not configured"

HCP Terraform couldn't reach Vault. Check:

1. The Vault URL in the test configuration is correct
2. The URL includes the protocol (`https://`) and port
3. Your network allows HCP Terraform to reach Vault
4. Vault's TLS certificate is valid (or Vault is configured to allow insecure connections)

### Error: "Role not found"

The specified role doesn't exist. Verify:

1. The role name in HCP Terraform matches the role in Vault exactly (case-sensitive)
2. The role exists in the correct namespace (if using Vault Enterprise)

> 3. The auth path is correct (default is `jwt`)

Reviewer: ```suggestion
The specified role doesn't exist. Verify the following conditions:

1. The role name in HCP Terraform matches the role in Vault exactly. Role names are case-sensitive.
1. If using Vault Enterprise, the role exists in the correct namespace.
1. The auth path is correct. The default path is `jwt`.
```
